### DISCUSSION TITLE
Energy with no relay

### DISCUSSION THREAD

**Discussion by rescalda** [arendst/Tasmota#14875] [Q&A] [answered]
Hello, may be a stupid question, easily I am loosing something very basic about Tasmota, but I really cannot understand what I am missing. I have a SMTONOFF and it works fine with Tasmota. I modified it while I don't want a Relay in that part of my circuit and I physically removed it, replaced with a correct size wire. Tasmota still works fine, obviously. While there is no relay, I removed the relay from the template and the related button. At this point, even if I see in the log the serial incoming data, the web interface reports all zeros. After some experimenting, I noticed I must have a relay and it must be ON. 
I issued the **PowerOnState 4** command and still I have all zeros. The only way to see the values, is to declare a relay, but this shows a useless button on the web interface. Luckily, thanks to PowerOnState 4, it has no effect.
Can I have only the CSE7766 in the template and have forever energy data displayed? 
Thanks.

**Comment by sfromis** [comment]
I'd try if [SetOption21](https://tasmota.github.io/docs/Commands/#setoption21) works for that.

**Reply by rescalda** [reply]
Yes, of course it works!
**SetOption21 | Energy monitoring when power is off**
The crazy thing is that I read **all** of them, the full _commands_ section, and I didn't notice this very clear statement!
Sorry.

**Reply by barbudor** [reply]
Do you really get power? No only voltage?
From my experience I was not supposed to happen

**Reply by rescalda** [reply]
Yes, with _SetOption21 1_ I get **all** the valuea!
![Meter](https://user-images.githubusercontent.com/38831927/154293326-6f5b077d-8291-492a-bbee-378a625bcdd3.jpg)

**Reply by rescalda** [reply]
By the way, a virtual relay on an unused pin is not a problem. What disturbed me was the useless button on the web interface. May be it could turned off? I don't know yet.

**Comment by barbudor** [comment, marked as answer]
While Setoption41 will allow reporting the voltage, it will not help with current.
The code as a test which force energy to 0 if power is not on.
The reason is that many power measurement circuit will still provide very small but non null values even if no load is connected and that would be obviously wrong. Accumulation over time would create an unacceptable drift

So you must keep a GPIO configured as a relay (virtual in your case) set to permanently ON with PowerOnState 4

**Reply by sfromis** [reply]
Suspecting that Tasmota may remember power state from when the pin was configured as a relay, I did a small test. Configured a gpio as a Relay, turned it on, before configuring to None. As expected no power state in the web UI. Then I configured it as a Relay again, and after the restart, the old state of "on" was restored in the web UI. As the state is remembered, it seems plausible that this is is what's tested by the code in the driver. Of course, this is merely speculation, but seems at least somewhat plausible.

**Reply by rescalda** [reply]
I hope this cannot lead to the loss of measuring after some events, for instance a power off!
Do you suggest to re-configure a dummy relay? Is there a way to suppress the related button in the web interface?

**Reply by sfromis** [reply]
I'd be inclined to have a dummy relay configured on an unused pin, ignoring the irrelevant button rendered, as the web UI is not really meant for anything but admin/test, with the "daily operation" expected to happen backend. You could use `WebButton` to make it "less inviting".

However, if my speculation about the power state being left after unconfiguring the pin is correct, I'd expect this to be pretty stable, as "normal events" would not "know" about it. It would probably be lost if you directly reset/restore the configuration, perhaps if you do major changes, or just possibly after a boot loop (should not occur). In case of it being lost rarely, a plan could be to temporarily config a dummy relay, and turn it on.

**Reply by barbudor** [reply]
I think that PowerOnState 4 disable the web button action to. Description says "disable further power control"

**Reply by sfromis** [reply]
Not really, if wanting to get rid of the button. Toggle is still where with `PowerOnState 4`. Just doesn't work.
